text/3855-mitigation-enforcement.md

```diff
@@ -0,0 +1,347 @@
+- Feature Name: `mitigation_enforcement`
+- Start Date: 2025-09-13
+- RFC PR: [rust-lang/rfcs#3855](https://github.com/rust-lang/rfcs/pull/3855)
+- Rust Issue: None
+
+# Summary
+[summary]: #summary
+
+Introduce the concept of "mitigation enforcement", so that when compiling
+a crate with mitigations enabled (for example, `-C stack-protector`),
+a compilation error will happen if the produced artifact would contain Rust
+code without the same mitigations enabled.
+
+This in many cases would require use of `-Z build-std`, since the standard
+library only comes with a single set of enabled mitigations per target.
+
+Mitigation enforcement should be disableable by the end-user via a compiler
+flag.
+
+# Motivation
+[motivation]: #motivation
+
+Memory unsafety mitigations are important for reducing the chance that a vulnerability
+ends up being exploitable.
+
+While in Rust, memory unsafety is less of a concern than in C, mitigations are
+still important for several reasons:
+
+1. Some mitigations (for example, straight line speculation mitigation,
+   [`-Z harden-sls`]) mitigate the impact of Spectre-style speculative
+   execution vulnerabilities, that exist in Rust just as well as C.
+2. Many Rust programs also contain large C/C++ components, that can have
+   memory vulnerabilities.
+3. Many Rust programs use `unsafe`, that can introduce memory unsafety
+   and vulnerabilities.
+
+Mitigations are generally enabled by passing a flag to the compiler (for
+example, [`-Z harden-sls`] or [`-Z stack-protector`]). If the compilation
```

Comment:
You always talk about stack-protector here, since that's your primary motivation of course, but I think it would be good to more explicitly list out all kinds of mitigations that Rust has it that people would like Rust to have in the future, to ensure that this makes sense for all of them (for example, the ones from https://doc.rust-lang.org/nightly/rustc/exploit-mitigations.html#exploit-mitigations-1).

I would especially be interested in whether there are *existing stable* mitigations that would like to make use of this, especially if it has a flag to toggle it.

Reply:
Some exploit mitigations that would benefit from it (e.g., CFI) precedes the Target Modifiers feature (which was intended to also solve this), but I don't think there are any stable exploit mitigations except maybe `-C control-flow-guard`.